[quant][graphmode] quant_fusion jit pass

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #25127 [quant][graphmode] insert_quant_dequant work with qconfig_dict
* #25069 [graphmode] insert_observers use qconfig_dict
* #25000 [cleanup] Remove InsertQuantDeQuantNode
* #24999 [cleanup] Remove insert_observers pass
* #24998 [cleanup] Remove deprecated graph mode quantization tests
* #24428 [quant][graphmode] Integration tests for initial quantization graph mode
* **#24427 [quant][graphmode] quant_fusion jit pass**
* #24426 [quant][graphmode] insert_quant_dequant jit pass

Summary:
Added following pass:
- _jit_pass_quant_fusion: Fusion pass that replaces the dequant->conv->quant patterns
to quantized_conv

Test Plan:
python test/test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D17001142](https://our.internmc.facebook.com/intern/diff/D17001142)

jerryzh168 committed:
[quant][graphmode] quant_fusion jit pass

Summary:
Added following pass:
- _jit_pass_quant_fusion: Fusion pass that replaces the dequant->conv->quant patterns
to quantized_conv

Test Plan:
python test/test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Update on "[quant][graphmode] quant_fusion jit pass"


Summary:
Added following pass:
- _jit_pass_quant_fusion: Fusion pass that replaces the dequant->conv->quant patterns
to quantized_conv

Test Plan:
python test/test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Update on "[quant][graphmode] quant_fusion jit pass"


Summary:
Added following pass:
- _jit_pass_quant_fusion: Fusion pass that replaces the dequant->conv->quant patterns
to quantized_conv

Test Plan:
python test/test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Update on "[quant][graphmode] quant_fusion jit pass"


Summary:
Added following pass:
- _jit_pass_quant_fusion: Fusion pass that replaces the dequant->conv->quant patterns
to quantized_conv

Test Plan:
python test/test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Update on "[quant][graphmode] quant_fusion jit pass"


Summary:
Added following pass:
- _jit_pass_quant_fusion: Fusion pass that replaces the dequant->conv->quant patterns
to quantized_conv

Test Plan:
python test/test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Update on "[quant][graphmode] quant_fusion jit pass"


Summary:
Added following pass:
- _jit_pass_quant_fusion: Fusion pass that replaces the dequant->conv->quant patterns
to quantized_conv

Test Plan:
python test/test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Update on "[quant][graphmode] quant_fusion jit pass"


Summary:
Added following pass:
- _jit_pass_quant_fusion: Fusion pass that replaces the dequant->conv->quant patterns
to quantized_conv

Test Plan:
python test/test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Update on "[quant][graphmode] quant_fusion jit pass"


Summary:
Added following pass:
- _jit_pass_quant_fusion: Fusion pass that replaces the dequant->conv->quant patterns
to quantized_conv

Test Plan:
python test/test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Update on "[quant][graphmode] quant_fusion jit pass"


Summary:
Added following pass:
- _jit_pass_quant_fusion: Fusion pass that replaces the dequant->conv->quant patterns
to quantized_conv

Test Plan:
python test/test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Update on "[quant][graphmode] quant_fusion jit pass"


Summary:
Added following pass:
- _jit_pass_quant_fusion: Fusion pass that replaces the dequant->conv->quant patterns
to quantized_conv

Test Plan:
python test/test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Update on "[quant][graphmode] quant_fusion jit pass"


Summary:
Added following pass:
- _jit_pass_quant_fusion: Fusion pass that replaces the dequant->conv->quant patterns
to quantized_conv

Test Plan:
python test/test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

nishantpdce commented:
What is the order? First we do fusion and then quant-dequant or otherway. 
Eg for conv->relu. can u explain

nishantpdce commented:
I think in future you will have more fusion patterns. How will the QuantFusion look like ? Should you not iterate over set of patterns ? I will let Mikhail provide more comments as he is better familiar with the fusion patterns

jerryzh168 commented:
> What is the order? First we do fusion and then quant-dequant or otherway.
> Eg for conv->relu. can u explain

conv->relu fusion will happen here as well, this PR is just used as a placeholder for quant_fusion pass, and we'll extend this later.

jerryzh168 committed:
Update on "[quant][graphmode] quant_fusion jit pass"


Summary:
Added following pass:
- _jit_pass_quant_fusion: Fusion pass that replaces the dequant->conv->quant patterns
to quantized_conv

Test Plan:
python test/test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

ZolotukhinM commented on test/test_jit.py:
```diff
@@ -1210,7 +1210,51 @@ def get_forward(m):
+        m = torch.jit.script(M())
+        observer = torch.jit.script(Observer())
+        torch._C._jit_pass_constant_propagation(get_forward(m).graph)
+        m._c = torch._C._jit_pass_prepare_quant(m._c, "forward",
```
We don't need to run all the previous passes to test this one. Just explicitly write out the IR we expect on the input and run only this pass on it. The test checking that all passes can work together is also needed, but it's not a unit test (and I think we already have such a test somewhere else, right?).

ZolotukhinM reviewed (commented):
@nishantpdce, when we have several patterns, they will be matched in order they are registered, so it's up to us to decide what we want first. In our case I'd expect the order will be from bigger to smaller, e.g:
1) DQ-Conv-Relu-Q --> QConvRelu
2) Conv-Relu --> ConvRelu

IOW, we would like to first check patterns that include Q/DQ before we check patterns from generic fusion.

ZolotukhinM reviewed (approved):
Overall looks good, some remarks are inline!

jerryzh168 committed:
Update on "[quant][graphmode] quant_fusion jit pass"


Summary:
Added following pass:
- _jit_pass_quant_fusion: Fusion pass that replaces the dequant->conv->quant patterns
to quantized_conv

Test Plan:
python test/test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Update on "[quant][graphmode] quant_fusion jit pass"


Summary:
Added following pass:
- _jit_pass_quant_fusion: Fusion pass that replaces the dequant->conv->quant patterns
to quantized_conv

Test Plan:
python test/test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Update on "[quant][graphmode] quant_fusion jit pass"


Summary:
Added following pass:
- _jit_pass_quant_fusion: Fusion pass that replaces the dequant->conv->quant patterns
to quantized_conv

Test Plan:
python test/test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Update on "[quant][graphmode] quant_fusion jit pass"


Summary:
Added following pass:
- _jit_pass_quant_fusion: Fusion pass that replaces the dequant->conv->quant patterns
to quantized_conv

Test Plan:
python test/test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -829,10 +829,9 @@ void quantizeTensor(const script::Module& module,
     const std::string& method_name) {
-  // TODO: Uncomment
-  //script::Module module = input_module.clone();
+  script::Module module = input_module.clone();
```
these changes are gone I think
